### PR TITLE
Add Baby Random Battle

### PR DESCRIPTION
**Pull request by adrivrie**
Adds a new format: Baby Random Battle! Design started about a year and a half ago, but it's finally ready.
Largely designed by [bobomania](https://www.smogon.com/forums/members/bobomania.553556/) with some help, implemented by me.

`sets.ts` is all new, though it follows the same format as most other rands sets. `teams.ts` is inherited from regular rands where possible, but a lot had to be changed. The name was decided by consensus from rands staff, to somewhat disassociate it from regular LC which has rules that the rands version doesn't follow.

It's tested as well as I was able to, so it's time for reviews!

### CODE REVIEW

File: data/random-battles/gen9baby/teams.ts [hunk 1]
```diff
@@ -0,0 +1,876 @@
...
+		const pokemon: RandomTeamsTypes.RandomSet[] = [];
+
+		// For Monotype
+		const isMonotype = !!this.forceMonotype || ruleTable.has('sametypeclause');
+		const typePool = this.dex.types.names().filter(name => name !== "Stellar");
+		const type = this.forceMonotype || this.sample(typePool);
+
+		const baseFormes: {[k: string]: number} = {};
+
+		const typeCount: {[k: string]: number} = {};
```
Comment (KrisXV) [inline data/random-battles/gen9baby/teams.ts:736]:
```suggestion
		const typeCount = new Utils.Multiset<string>();
```

File: data/random-battles/gen9baby/teams.ts [hunk 2]
```diff
@@ -0,0 +1,876 @@
...
+			);
+			// Dynamically scale limits for different team sizes. The default and minimum value is 1.
+			const limitFactor = Math.round(this.maxTeamSize / 6) || 1;
+
+			if (!isMonotype && !this.forceMonotype) {
+				let skip = false;
+
+				// Limit two of any type
+				for (const typeName of types) {
+					if (typeCount[typeName] >= 2 * limitFactor) {
```
Comment (KrisXV) [inline data/random-battles/gen9baby/teams.ts:770]:
```suggestion
					if ((typeCount.get(typeName) || 0) >= 2 * limitFactor) {
```

File: data/random-battles/gen9baby/teams.ts [hunk 3]
```diff
@@ -0,0 +1,876 @@
...
+			// Now that our Pokemon has passed all checks, we can increment our counters
+			baseFormes[species.baseSpecies] = 1;
+
+			// Increment type counters
+			for (const typeName of types) {
+				if (typeName in typeCount) {
+					typeCount[typeName]++;
+				} else {
+					typeCount[typeName] = 1;
+				}
```
Comment (KrisXV) [inline data/random-battles/gen9baby/teams.ts:823]:
```suggestion
				typeCount.add(typeName);
```

File: data/random-battles/gen9baby/teams.ts [hunk 4]
```diff
@@ -0,0 +1,876 @@
...
+
+		// For Monotype
+		const isMonotype = !!this.forceMonotype || ruleTable.has('sametypeclause');
+		const typePool = this.dex.types.names().filter(name => name !== "Stellar");
+		const type = this.forceMonotype || this.sample(typePool);
+
+		const baseFormes: {[k: string]: number} = {};
+
+		const typeCount: {[k: string]: number} = {};
+		const typeComboCount: {[k: string]: number} = {};
```
Comment (KrisXV) [inline data/random-battles/gen9baby/teams.ts:737]:
```suggestion
		const typeComboCount = new Utils.Multiset<string>();
```

File: data/random-battles/gen9baby/teams.ts [hunk 5]
```diff
@@ -0,0 +1,876 @@
...
+
+				// Limit four weak to Freeze-Dry
+				if (weakToFreezeDry) {
+					if (!typeWeaknesses['Freeze-Dry']) typeWeaknesses['Freeze-Dry'] = 0;
+					if (typeWeaknesses['Freeze-Dry'] >= 4 * limitFactor) continue;
+				}
+			}
+
+			// Limit three of any type combination in Monotype
+			if (!this.forceMonotype && isMonotype && (typeComboCount[typeCombo] >= 3 * limitFactor)) continue;
```
Comment (KrisXV) [inline data/random-battles/gen9baby/teams.ts:805]:
```suggestion
			if (!this.forceMonotype && isMonotype && ((typeComboCount.get(typeCombo) || 0) >= 3 * limitFactor)) continue;
```

File: data/random-battles/gen9baby/teams.ts [hunk 6]
```diff
@@ -0,0 +1,876 @@
...
+					typeCount[typeName]++;
+				} else {
+					typeCount[typeName] = 1;
+				}
+			}
+			if (typeCombo in typeComboCount) {
+				typeComboCount[typeCombo]++;
+			} else {
+				typeComboCount[typeCombo] = 1;
+			}
```
Comment (KrisXV) [inline data/random-battles/gen9baby/teams.ts:829]:
```suggestion
			typeComboCount.add(typeCombo);
```

File: data/random-battles/gen9baby/teams.ts [hunk 7]
```diff
@@ -0,0 +1,876 @@
...
+		// For Monotype
+		const isMonotype = !!this.forceMonotype || ruleTable.has('sametypeclause');
+		const typePool = this.dex.types.names().filter(name => name !== "Stellar");
+		const type = this.forceMonotype || this.sample(typePool);
+
+		const baseFormes: {[k: string]: number} = {};
+
+		const typeCount: {[k: string]: number} = {};
+		const typeComboCount: {[k: string]: number} = {};
+		const typeWeaknesses: {[k: string]: number} = {};
```
Comment (KrisXV) [inline data/random-battles/gen9baby/teams.ts:738]:
```suggestion
		const typeWeaknesses = new Utils.Multiset<string>();
```

File: data/random-battles/gen9baby/teams.ts [hunk 8]
```diff
@@ -0,0 +1,876 @@
...
+					}
+				}
+				if (skip) continue;
+
+				// Limit three weak to any type, and one double weak to any type
+				for (const typeName of this.dex.types.names()) {
+					// it's weak to the type
+					if (this.dex.getEffectiveness(typeName, species) > 0) {
+						if (!typeWeaknesses[typeName]) typeWeaknesses[typeName] = 0;
+						if (typeWeaknesses[typeName] >= 3 * limitFactor) {
```
Comment (KrisXV) [inline data/random-battles/gen9baby/teams.ts:782]:
```suggestion
						if ((typeWeaknesses.get(typeName) || 0) >= 3 * limitFactor) {
```

File: data/random-battles/gen9baby/teams.ts [hunk 9]
```diff
@@ -0,0 +1,876 @@
...
+							break;
+						}
+					}
+				}
+				if (skip) continue;
+
+				// Limit four weak to Freeze-Dry
+				if (weakToFreezeDry) {
+					if (!typeWeaknesses['Freeze-Dry']) typeWeaknesses['Freeze-Dry'] = 0;
+					if (typeWeaknesses['Freeze-Dry'] >= 4 * limitFactor) continue;
```
Comment (KrisXV) [inline data/random-battles/gen9baby/teams.ts:800]:
```suggestion
					if ((typeWeaknesses.get('Freeze-Dry') || 0) >= 4 * limitFactor) continue;
```

File: data/random-battles/gen9baby/teams.ts [hunk 10]
```diff
@@ -0,0 +1,876 @@
...
+				typeComboCount[typeCombo]++;
+			} else {
+				typeComboCount[typeCombo] = 1;
+			}
+
+			// Increment weakness counter
+			for (const typeName of this.dex.types.names()) {
+				// it's weak to the type
+				if (this.dex.getEffectiveness(typeName, species) > 0) {
+					typeWeaknesses[typeName]++;
```
Comment (KrisXV) [inline data/random-battles/gen9baby/teams.ts:835]:
```suggestion
					typeWeaknesses.add(typeName);
```

File: data/random-battles/gen9baby/teams.ts [hunk 11]
```diff
@@ -0,0 +1,876 @@
...
+				}
+				if (this.dex.getEffectiveness(typeName, species) > 1) {
+					typeDoubleWeaknesses[typeName]++;
+				}
+			}
+			if (weakToFreezeDry) typeWeaknesses['Freeze-Dry']++;
```
Comment (KrisXV) [inline data/random-battles/gen9baby/teams.ts:841]:
```suggestion
			if (weakToFreezeDry) typeWeaknesses.add('Freeze-Dry');
```

File: data/random-battles/gen9baby/teams.ts [hunk 12]
```diff
@@ -0,0 +1,876 @@
...
+		const isMonotype = !!this.forceMonotype || ruleTable.has('sametypeclause');
+		const typePool = this.dex.types.names().filter(name => name !== "Stellar");
+		const type = this.forceMonotype || this.sample(typePool);
+
+		const baseFormes: {[k: string]: number} = {};
+
+		const typeCount: {[k: string]: number} = {};
+		const typeComboCount: {[k: string]: number} = {};
+		const typeWeaknesses: {[k: string]: number} = {};
+		const typeDoubleWeaknesses: {[k: string]: number} = {};
```
Comment (KrisXV) [inline data/random-battles/gen9baby/teams.ts:739]:
```suggestion
		const typeDoubleWeaknesses = new Utils.Multiset<string>();
```

File: data/random-battles/gen9baby/teams.ts [hunk 13]
```diff
@@ -0,0 +1,876 @@
...
+					if (this.dex.getEffectiveness(typeName, species) > 0) {
+						if (!typeWeaknesses[typeName]) typeWeaknesses[typeName] = 0;
+						if (typeWeaknesses[typeName] >= 3 * limitFactor) {
+							skip = true;
+							break;
+						}
+					}
+					if (this.dex.getEffectiveness(typeName, species) > 1) {
+						if (!typeDoubleWeaknesses[typeName]) typeDoubleWeaknesses[typeName] = 0;
+						if (typeDoubleWeaknesses[typeName] >= 1 * limitFactor) {
```
Comment (KrisXV) [inline data/random-battles/gen9baby/teams.ts:789]:
```suggestion
						if ((typeDoubleWeaknesses.get(typeName) || 0) >= 1 * limitFactor) {
```

File: data/random-battles/gen9baby/teams.ts [hunk 14]
```diff
@@ -0,0 +1,876 @@
...
+			}
+
+			// Increment weakness counter
+			for (const typeName of this.dex.types.names()) {
+				// it's weak to the type
+				if (this.dex.getEffectiveness(typeName, species) > 0) {
+					typeWeaknesses[typeName]++;
+				}
+				if (this.dex.getEffectiveness(typeName, species) > 1) {
+					typeDoubleWeaknesses[typeName]++;
```
Comment (KrisXV) [inline data/random-battles/gen9baby/teams.ts:838]:
```suggestion
					typeDoubleWeaknesses.add(typeName);
```

File: data/random-battles/gen9baby/teams.ts [hunk 15]
```diff
@@ -0,0 +1,876 @@
...
+				}
+			}
+			if (weakToFreezeDry) typeWeaknesses['Freeze-Dry']++;
+
+			// Track what the team has
+			if (set.ability === 'Drizzle' || set.moves.includes('raindance')) teamDetails.rain = 1;
+			if (set.ability === 'Drought' || set.moves.includes('sunnyday')) {
+				teamDetails.sun = 1;
+			}
```
Comment (KrisXV) [inline data/random-battles/gen9baby/teams.ts:847]:
```suggestion
			if (set.ability === 'Drought' || set.moves.includes('sunnyday')) teamDetails.sun = 1;
```

**Comment by adrivrie** [comment]
Thanks for the review @KrisXV! I camelCased the things I forgot.
The other stuff was mostly in code copied from the "main" `teams.ts` in /gen9/, so that file might itself need a cleanup at some point

**Pull request merged**